### DISCUSSION TITLE
Extending Jackson JsonTypeInfo support to consider defaultImpl

### DISCUSSION THREAD

**Discussion by eleftherias** [victools/jsonschema-generator#278] [Ideas]
The Jackson `@JsonTypeInfo` annotation has a property called `defaultImpl`.
From the `defaultImpl` Javadoc:
> Optional property that can be used to specify default implementation class to use for deserialization if type identifier is either not present,  or can not be mapped to a registered type.

Example usage:
```java
@JsonTypeInfo(use = Id.NAME, property = "type", defaultImpl = Dog.class)
````

I imagine this project could support `defaultImpl`, by marking the Jackson subtype discriminator as non-required for the class in  `defaultImpl`. 

In the example above, the generated schema would contain:
```JSON
    "Dog-2" : {
      "allOf" : [ {
        "$ref" : "#/definitions/Dog-1"
      }, {
        "type" : "object",
        "properties" : {
          "type" : {
            "const" : "Dog"
          }
        }
      } ]
    },
    "Cat-2" : {
      "allOf" : [ {
        "$ref" : "#/definitions/Cat-1"
      }, {
        "type" : "object",
        "properties" : {
          "type" : {
            "const" : "Cat"
          }
        },
        "required" : [ "type" ]
      } ]
    },
    "Mouse-2" : {
      "allOf" : [ {
        "$ref" : "#/definitions/Mouse-1"
      }, {
        "type" : "object",
        "properties" : {
          "type" : {
            "const" : "Mouse"
          }
        },
        "required" : [ "type" ]
      } ]
    },
```

Notice that "type" is not required for Dog-2, but is still required for Cat-2 and Mouse-2.

This way if "type" is missing, it defaults to "Dog", which is what `defaultImpl` describes.

**Comment by CarstenWickner** [comment]
The suggestion sounds good, and I've created PR #279 for that now.
It does cover only the first part though:
> if type identifier is either not present,

For the second part, one would have to list all known identifiers and negate them:
> or can not be mapped to a registered type.

While possible, I'd refrain from supporting this second half for now, as it seems rather excessive. If included, that should probably be hidden behind a new `JacksonOption`then, to enable easily toggling it on/off.

Or what do you reckon @eleftherias?

**Comment by eleftherias** [comment]
Thanks @CarstenWickner!
I agree that supporting the case where the type identifier is not present is adequate.

**Comment by CarstenWickner** [comment]
The change has been released as [v4.26.0](https://github.com/victools/jsonschema-generator/releases/tag/v4.26.0).
